src/Neo/Cryptography/Crypto.cs

```diff
@@ -87,24 +96,35 @@ public static byte[] Sign(byte[] message, byte[] priKey, ECC.ECCurve ecCurve = n
                 Curve = curve,
                 D = priKey,
             });
-            return ecdsa.SignData(message, HashAlgorithmName.SHA256);
+            var hashAlg =
+                hasher == Hasher.SHA256 ? HashAlgorithmName.SHA256 :
+                throw new NotSupportedException(nameof(hasher));
+            return ecdsa.SignData(message, hashAlg);
         }
 
         /// <summary>
-        /// Verifies that a digital signature is appropriate for the provided key and message.
+        /// Verifies that a digital signature is appropriate for the provided key, message and hash algorithm.
         /// </summary>
         /// <param name="message">The signed message.</param>
         /// <param name="signature">The signature to be verified.</param>
         /// <param name="pubkey">The public key to be used.</param>
+        /// <param name="hasher">The hash algorithm to be used to hash the message, the default is SHA256.</param>
         /// <returns><see langword="true"/> if the signature is valid; otherwise, <see langword="false"/>.</returns>
-        public static bool VerifySignature(ReadOnlySpan<byte> message, ReadOnlySpan<byte> signature, ECC.ECPoint pubkey)
+        public static bool VerifySignature(ReadOnlySpan<byte> message, ReadOnlySpan<byte> signature, ECC.ECPoint pubkey, Hasher hasher = Hasher.SHA256)
         {
             if (signature.Length != 64) return false;
 
-            if (IsOSX && pubkey.Curve == ECC.ECCurve.Secp256k1)
+            if (hasher == Hasher.Keccak256 || (IsOSX && pubkey.Curve == ECC.ECCurve.Secp256k1))
```

Comment:
should the else branch be dropped?

Reply:
No, because we still have to process SHA256 with Koblitz curve (if not on mac).

Reply:
but i think BouncyCastle is able to handle this case, isn't it?

Reply:
maybe the `IsOSX && pubkey.Curve == ECC.ECCurve.Secp256k1` is introduced here for performance optimization, @shargon do you have some comments here?

Reply:
`Secp256k1` isn't natively supported on `MacOS`; reason we need to use `BouncyCastle`.

Reply:
@cschuchardt88 why not just use BouncyCastle and drop the **else** branch

Reply:
In general, to me it seems to be more preferable (and safe) to use native built-in implementation instead of some external libraries, and thus, we're stick to the native ECDsa's sign/verify implementation as much as possible. In future both k1 on macOS and Koblitz hasher may be added to the native library, then we may drop this `if`. But until then I think it's OK to keep BouncyCastle for these two cases.

Reply:
I'm not 100% sure why. Maybe we want to avoid 3rd-party libraries or unsupported features for linux/windows support.

Maybe @shargon @Jim8y can help us.

Reply:
> In general, to me it seems to be more preferable (and safe) to use native built-in implementation instead of some external libraries, and thus, we're stick to the native ECDsa's sign/verify implementation as much as possible. In future both k1 on macOS and Koblitz hasher may be added to the native library, then we may drop this `if`. But until then I think it's OK to keep BouncyCastle for these two cases.

I hold the opposite view

for blockchain, consistency is very important, and behavioral differences caused by architecture and platform cannot be tolerated.

the implementation of dotnet's Crypto library is to directly call the operating system API instead of implementing it by themselves. even their own developers are not clear enough about these behaviors.

Reply:
Let's then start a discussion issue, https://github.com/neo-project/neo/issues/3220.